### PR TITLE
Teradata JDBC Extractor and Source

### PR DESCRIPTION
**Pull request by ypopov**
Do not required extended access rights  - schema generated from generic JDBC ResultSetMetaData. Works with views and custom queries  (use source.querybased.is.metadata.column.check.enabled=false for derived columns). Tested against Teradata version 13 and 15.

### CODE REVIEW

File: gobblin-core/src/main/java/gobblin/source/extractor/extract/jdbc/TeradataExtractor.java [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+      List<Predicate> predicateList) throws IOException {
+    return null;
+  }
+
+  @Override
+  public String getConnectionUrl() {
+    String host = this.workUnit.getProp(ConfigurationKeys.SOURCE_CONN_HOST_NAME);
+    String port = this.workUnit.getProp(ConfigurationKeys.SOURCE_CONN_PORT,"1025");
+    String database = this.workUnit.getProp(ConfigurationKeys.SOURCE_QUERYBASED_SCHEMA);
+    String url = "jdbc:teradata://" + host.trim() + "/DATABASE=" + database.trim() + ",DBS_PORT=" + port.trim() ;
```
Comment (lbendig) [inline gobblin-core/src/main/java/gobblin/source/extractor/extract/jdbc/TeradataExtractor.java]: When using the Teradata JDBC driver the `TYPE=FASTEXPORT` parameter could be added to the connection string. Optionally a default charset as well (e.g: `CHARSET=utf16`).

File: gobblin-core/src/main/java/gobblin/source/extractor/extract/jdbc/TeradataExtractor.java [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+            .put("varchar", "string")
+            .put("byte", "bytes")
+            .put("varbyte", "bytes")
+            .put("date", "date")
+            .put("time", "time")
+            .put("timestamp", "timestamp")
+            .put("clob", "string")
+            .put("blob", "string")
+            .put("structured udt", "array")
+            .build();
```
Comment (lbendig) [inline gobblin-core/src/main/java/gobblin/source/extractor/extract/jdbc/TeradataExtractor.java]: In our version we have these additional mappings:
```
.put("double precision", "float")
.put("numeric", "double")
.put("real", "float")
.put("character", "string")
.put("char varying", "string")
.put("character varying", "string")
.put("long varchar", "string")
.put("interval", "string")
```

File: gobblin-core/src/main/java/gobblin/source/extractor/extract/jdbc/TeradataExtractor.java [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+    if (isNullOrEmpty(watermarkFilter)) {
+      watermarkFilter = "1=1";
+    }
+    query = query.replace(this.getOutputColumnProjection(), columnProjection)
+        .replace(ConfigurationKeys.DEFAULT_SOURCE_QUERYBASED_WATERMARK_PREDICATE_SYMBOL, watermarkFilter);
+    String sampleFilter = this.constructSampleClause();
+    query = query + sampleFilter;
+
+    if (!isNullOrEmpty(sampleFilter)) {
+      query = "SELECT COUNT(1) FROM (" + query.replace(" COUNT(1) ", " 1 at t") + ") temp";
```
Comment (lbendig) [inline gobblin-core/src/main/java/gobblin/source/extractor/extract/jdbc/TeradataExtractor.java]: Teradata count returns an `INT`. For big tables (rows > 2^31-1) this query produces a numerical overflow.
I'd suggest to have `CAST(COUNT(1) AS BIGINT)` instead.

File: gobblin-core/src/main/java/gobblin/source/extractor/extract/jdbc/TeradataExtractor.java [hunk 4]
```diff
@@ -188,12 +192,15 @@ public TeradataExtractor(WorkUnitState workUnitState) {
...
+    String urlPrefix = "jdbc:teradata://"; 
     String host = this.workUnit.getProp(ConfigurationKeys.SOURCE_CONN_HOST_NAME);
+    checkArgument(!isNullOrEmpty(host), "Connectionn host cannot be null or empty at %s", ConfigurationKeys.SOURCE_CONN_HOST_NAME);
+    
     String port = this.workUnit.getProp(ConfigurationKeys.SOURCE_CONN_PORT,"1025");
     String database = this.workUnit.getProp(ConfigurationKeys.SOURCE_QUERYBASED_SCHEMA);
-    String url = "jdbc:teradata://" + host.trim() + "/DATABASE=" + database.trim() + ",DBS_PORT=" + port.trim() ;
-    
-    return url;
+    String defaultUrl = urlPrefix + host.trim() + "/DATABASE=" + database.trim() + ",DBS_PORT=" + port.trim() ;
```
Comment (lbendig) [inline gobblin-core/src/main/java/gobblin/source/extractor/extract/jdbc/TeradataExtractor.java]: I think it's worth to add the `TYPE=FASTEXPORT` connection parameter by default.

**Comment by lbendig** [comment]
Tested the `TeradataExtractor` for full dumps, custom watermark ranges, custom queries and sampling, looks good!

**Comment by chavdar** [comment]
LGTM.

**Pull request merged**